src/leader_scheduler.rs

```diff
@@ -0,0 +1,776 @@
+    pub bootstrap_leader: Pubkey,
+
+    // The last height at which the bootstrap_leader will be in power before
+    // the leader rotation process begins to pick future leaders
+    pub bootstrap_height: u64,
+
+    // Maintain the set of active validators
+    pub active_validators: ActiveValidators,
+
+    // Round-robin ordering for the validators
+    leader_schedule: Vec<Pubkey>,
+
+    // The last height at which the seed + schedule was generated
+    last_seed_height: Option<u64>,
+
+    // The seed used to determine the round robin order of leaders
+    seed: u64,
+}
+
+// The LeaderScheduler implements a schedule for leaders as follows:
+//
+// 1) During the bootstrapping period of bootstrap_height PoH counts, the
+// leader is hard-coded to the input bootstrap_leader.
+//
+// 2) After the first seed is generated, this signals the beginning of actual leader rotation.
+// From this point on, every seed_rotation_interval PoH counts we generate the seed based
+// on the PoH height, and use it to do a weighted sample from the set
+// of validators based on current stake weight. This gets you the first leader A for
+// the next leader_rotation_interval PoH counts. On the same PoH count we generate the seed,
+// we also order the validators based on their current stake weight, and starting
+// from leader A, we then pick the next leader sequentially every leader_rotation_interval
+// PoH counts based on this fixed ordering, so the next
+// seed_rotation_interval / leader_rotation_interval leaders are determined.
+//
+// 3) When we we hit the next seed rotation PoH height, step 2) is executed again to
+// calculate the leader schedule for the upcoming seed_rotation_interval PoH counts.
+impl LeaderScheduler {
+    pub fn new(config: &LeaderSchedulerConfig) -> Self {
+        let mut bootstrap_height = 1000;
+        if let Some(input) = config.bootstrap_height_option {
+            bootstrap_height = input;
+        }
+
+        let mut leader_rotation_interval = 100;
+        if let Some(input) = config.leader_rotation_interval_option {
+            leader_rotation_interval = input;
+        }
+
+        let mut seed_rotation_interval = 1000;
+        if let Some(input) = config.seed_rotation_interval_option {
+            seed_rotation_interval = input;
+        }
+
+        // Enforced invariants
+        assert!(seed_rotation_interval >= leader_rotation_interval);
+        assert!(bootstrap_height > 0);
+        assert!(seed_rotation_interval % leader_rotation_interval == 0);
+
+        LeaderScheduler {
+            active_validators: ActiveValidators::new(config.active_window_length_option),
+            leader_rotation_interval,
+            seed_rotation_interval,
+            leader_schedule: Vec::new(),
+            last_seed_height: None,
+            bootstrap_leader: config.bootstrap_leader,
+            bootstrap_height,
+            seed: 0,
+        }
+    }
+
+    pub fn push_vote(&mut self, id: Pubkey, height: u64) {
```

Comment:
is there a case where push_vote() does not imply update_height()?

Reply:
I guess we want to keep these separate so that there's just one place where we update height...